janitor/functions/case_when.py

```diff
@@ -128,71 +156,79 @@ def case_when(df: pd.DataFrame, *args, column_name: str) -> pd.DataFrame:
                 f"condition{index} and value{index} failed to evaluate. "
                 f"Original error message: {e}"
             ) from e
-
-    return df.assign(**{column_name: default})
+    return default
 
 
-def _case_when_checks(df: pd.DataFrame, args, column_name):
+def _case_when_checks(
+    df: pd.DataFrame, args, default
+) -> tuple[list, list, pd.Series]:
     """
     Preliminary checks on the case_when function.
     """
-    if len(args) < 3:
+    if len(args) < 2:
         raise ValueError(
-            "At least three arguments are required for the `args` parameter."
+            "At least two arguments are required for the `args` parameter"
         )
-    if len(args) % 2 != 1:
-        raise ValueError(
-            "It seems the `default` argument is missing from the variable "
-            "`args` parameter."
-        )
-
-    check("column_name", column_name, [str])
-
-    *args, default = args
 
     booleans = []
     replacements = []
     for index, value in enumerate(args):
-        if index % 2 == 0:
-            booleans.append(value)
-        else:
+        if index % 2:
             replacements.append(value)
+        else:
+            booleans.append(value)
+    if len(booleans) != len(replacements):
+        raise ValueError(
+            "The number of conditions and values do not match. "
+            f"There are {len(booleans)} conditions and {len(replacements)} "
+            "values."
+        )
 
-    conditions = []
-    for condition in booleans:
-        if callable(condition):
-            condition = apply_if_callable(condition, df)
-        elif isinstance(condition, str):
-            condition = df.eval(condition)
-        conditions.append(condition)
-
-    targets = []
-    for replacement in replacements:
-        if callable(replacement):
-            replacement = apply_if_callable(replacement, df)
-        targets.append(replacement)
+    booleans = [
+        apply_if_callable(condition, df)
+        if callable(condition)
+        else df.eval(condition)
+        if isinstance(condition, str)
+        else condition
+        for condition in booleans
+    ]
+
+    replacements = [
+        apply_if_callable(replacement, df)
+        if callable(replacement)
+        else replacement
+        for replacement in replacements
+    ]
 
     if callable(default):
         default = apply_if_callable(default, df)
-    if not is_list_like(default):
+    if pd.api.types.is_scalar(default):
         default = pd.Series([default]).repeat(len(df))
-        default.index = df.index
-    if not hasattr(default, "shape"):
-        default = pd.Series([*default])
+    if not is_array_like(default):
+        raise TypeError(
+            "The argument for the `default` parameter "
+            "should evaluate to an array-like object, "
+            f"instead got {type(default)}"
```

Comment:
```suggestion
            f"instead got {type(default)!r}"
```

This could output with `''`